### ISSUE TITLE
Images are not downloaded completely

### ISSUE THREAD

**Issue by wintstar** [linuxmint/cinnamon-spices-applets#4651]
@bing-wallpaper@starcross.dev

```
 "last-edited": 1664004583
Cinnamon 5.6.7
LMDE 5 Elsie
Intel 2nd Generation Core Processor Family Integrated Graphics / driver: i915 v: kernel bus ID: 00:02.0 chip ID: 8086:0126
Lenovo ThinkPad T420
Intel Core i5-2520M (2.50GHz, 3MB L3, 1333MHz FSB)
Intel HD 3000 Grafik
RAM 8GB

 ```

[
](https://github.com/linuxmint/cinnamon-spices-applets/tree/master/bing-wallpaper@starcross.dev)
**Notify author of applet**
@Starcross
@fredcw

Images are not downloaded completely. The bottom half of the image is missing.

It would be advantageous if the image of the day is downloaded after the start of the operating system with a time delay of 20 seconds. 

1) How can I reload or refresh the image if it was not loaded correctly?
2) How to change the location of the images (https://github.com/linuxmint/cinnamon-spices-applets/blob/master/bing-wallpaper%40starcross.dev/files/bing-wallpaper%40starcross.dev/applet.js#L41) instead of $user/.config/bingwallpaper to usr/share/backgrounds?

**Comment by Starcross** [comment]
1. There's no way to do this at the moment. I don't think it has ever happened to me, but I have heard from a few folks saying they have incomplete downloads. I'd prefer to be able to detect this automatically instead of have a manual process. 
2. The location is hard coded, however I think /usr/share/backgrounds requires root access? Only one file is stored anyway (the current one), so you won't be getting an archive of previous images of the day. There are plenty of website that publish these though

**Comment by wintstar** [comment]
This is the picture from today

Picture from today:
![BingWallpaper](https://user-images.githubusercontent.com/2753863/214801968-e84ab165-d33a-44b8-9ca5-e82c7349075c.jpg)

meta.json from today
`{"market":{"mkt":"de-DE"},"images":[{"startdate":"20230125","fullstartdate":"202301252300","enddate":"20230126","url":"/th?id=OHR.HighArchChina_DE-DE5049467350_1920x1080.jpg&rf=LaDigue_1920x1080.jpg&pid=hp","urlbase":"/th?id=OHR.HighArchChina_DE-DE5049467350","copyright":"Das Himmelstor im Tianmen-Berg-Nationalpark, China (© Shane P. White/Minden Pictures)","copyrightlink":"https://www.bing.com/search?q=Das+Himmelstor+im+Tianmen-Berg-Nationalpark+china&form=hpcapt&filters=HpDate%3a%2220230125_2300%22","title":"Treppe zum Himmel","quiz":"/search?q=Bing+homepage+quiz&filters=WQOskey:%22HPQuiz_20230125_HighArchChina%22&FORM=HPQUIZ","wp":true,"hsh":"a3529987450e8383f5805f02a4f59885","drk":1,"top":1,"bot":1,"hs":[]}],"tooltips":{"loading":"Wird geladen...","previous":"Vorheriges Bild","next":"Nächstes Bild","walle":"Dieses Bild steht nicht zum Download als Hintergrundbild zur Verfügung.","walls":"Dieses Bild herunterladen. Dieses Bild darf nur als Hintergrundbild verwendet werden."}}`

I think that when other services like Nextcloud sync client etc. are finished and then only the image is downloaded it is then also completely downloaded.

As you can see, you don't see a picture. It has not been downloaded properly. Probably because the download time was too short. I myself am familiar with Bing images. Created an application myself for my website [www.stephan-frank.de](https://www.stephan-frank.de/blog/im-reich-der-farben).

**Comment by Starcross** [comment]
My suggestion would be once your network connection is stable, delete the image and it should be automatically downloaded again within 5 minutes. It sounds like the download is not completing for some reason.

I know that with no internet connection, the file is not downloaded and it will try again every 5 minutes, successfully downloading once it becomes available. It sounds like in your case you are getting a partial file (you could check this) or an empty file (but there is a check for a file size of zero in place already)

**Comment by Drugwash2** [comment]
@Starcross Can't you use the hash value in **meta.json** to check for downloaded image's completion and if mismatched delete it and redownload within a specific time frame...? Also add a limited/configurable number or retries.

**Comment by wintstar** [comment]
The Picture from today
![BingWallpaper](https://user-images.githubusercontent.com/2753863/215260490-230b0211-61b1-4527-8075-ffe1e8f76309.jpg)

**Comment by Starcross** [comment]
Thanks, it's clearly an incomplete download.

I agree with @Drugwash2 regarding using the hash, however I can't find what hashing algorithm Microsoft is using here. Does anyone know?

**Comment by Drugwash2** [comment]
Alas, I've searched and computed a bunch of algorithms but none matched the one in the **meta.json** published above for the respective image. It _looks_ like a 128-bit hash but it may be salted with an unknown key. Darn, it woud've been too easy.

**Comment by wintstar** [comment]
I think it would be better and easier not to download the image directly at the start of the operating system. A time lock would have the advantage to let other programs do their work first and only then download the image.

**Comment by Drugwash2** [comment]
That would be guesswork of the worse kind than checking file validity as an image. Nobody can predict how long would a system take to fully load and when other applications/scripts/services/etc would be finished with whatever online connections and/or resources-heavy activities they may be running right at startup or on a timer.

Of course, a delay timer might help somewhat, but it won't be reliable, and if/when the applet tries to download an image in the middle of a session during heavy Internet traffic the same issue may happen again.

**Comment by Starcross** [comment]
I think the obvious thing to try would be get the declared `content-length` in the header and check if the received data matches that size, especially since this issue seems to be due to incomplete downloads. 

WRT when to download, I think the general expectation is that the image is downloaded as soon as possible on startup, hence you see today's 'image of the day' for the day! If the connection is unavailable or unstable it should periodically retry (which it does now)

**Comment by Drugwash2** [comment]
Yes, checking file length could be a valid action too - unless for some reason the size is reserved ahead and then filled up only to a certain point. No idea if this is the case but just thought it should be mentioned. Probably the best action in this case would be to first check file size against content-length and then check the presence of the JPEG end signature in the downloaded file.

**Comment by wintstar** [comment]
With commit https://github.com/linuxmint/cinnamon-spices-applets/commit/394ba2ff1075cc19a47478111e12e35076ce892b is the issue solved. Many thanks for update. This issue can closed.
